mesop/server/state_session.py

```diff
@@ -0,0 +1,91 @@
+from datetime import datetime, timedelta
+from typing import Any, Protocol
+
+from mesop.server.config import Config, app_config
+
+States = dict[type[Any], object]
+
+
+class StateSessionBackend(Protocol):
+  """Interface for state session backends."""
+
+  def restore(self, token: str, states: States):
+    """Gets the saved state from the given token and updates the given state."""
+    raise NotImplementedError()
+
+  def save(self, token: str, states: States):
+    """Saves state to the backend with the given token."""
+    raise NotImplementedError()
+
+  def clear_stale_sessions(self):
+    """Deletes unused state data."""
+    raise NotImplementedError()
+
+
+class NullStateSessionBackend(StateSessionBackend):
+  """State session backend used when state sessions are disabled."""
+
+  def restore(self, token: str, states: States):
+    raise Exception("No state session backend configured.")
+
+  def save(self, token: str, states: States):
+    pass
+
+  def clear_stale_sessions(self):
+    pass
+
+
+class MemoryStateSessionBackend(StateSessionBackend):
+  """Stores state in memory.
+
+  The main limitation of this backend is that deloyments with multiple processes /
+  replicas / instances will cause a lot of cache misses if there is no session affinity.
+
+  This is because memory is local to each Mesop server process.
+
```

Comment:
I'm wondering if we should check how much memory is left. Otherwise, if there's a bunch of sessions with large states (e.g. 100MB dataframe), will the server hit out of memory errors (OOM)?

Could use a library like:
https://github.com/giampaolo/psutil?tab=readme-ov-file#memory

Reply:
[By default, Cloud run has 512MB](https://cloud.google.com/run/docs/configuring/services/memory-limits#setting) so I think we need to be fairly conservative with how much memory we allocate for this

Reply:
Interestingly, Streamlit has a similar issue: https://github.com/streamlit/streamlit/issues/6602 IIUC, they run into this because their backend is always stateful.

Reply:
Yeah, memory capacity is definitely a concern with python. Even for normal sized objects, Python dicts tend can take up a fair amount of space. Definitely easy for memory to balloon. I think dataclasses are better in that regard. I think with dataclasses if you enable slots, that can also improve memory performance since with slots you can't dynamically add fields to the dataclass which I think we should restrict anyway.

This could be one reason serializing the data into strings even in the memory case could be beneficial. That could reduce the memory load a bit with the trade off of needing to serialize/deserialize each time. But I don't think it would help that much if users are making a lot of interactions per session. But could help with the stale session case.

-----

I like the idea of checking memory. Not ideal but at least we can fall back to sending the request over temporarily, but could lead to uneven performance for users. But better than getting OOM'd.

-----

Ideally, for real use cases, we would have a way to avoid using the memory backend.